compiler/src/dotty/tools/dotc/typer/Applications.scala

```diff
@@ -2133,8 +2133,45 @@ trait Applications extends Compatibility {
     }
   }
 
-  def isApplicableExtensionMethod(ref: TermRef, receiver: Type)(using Context) =
-    ref.symbol.is(ExtensionMethod)
-    && !receiver.isBottomType
-    && isApplicableMethodRef(ref, receiver :: Nil, WildcardType)
+  /** Assuming methodRef is a reference to an extension method defined e.g. as
+   *
+   *  extension [T1, T2](using A)(using B, C)(receiver: R)(using D)
+   *    def foo[T3](using E)(f: F): G = ???
+   *
+   *  return the tree representing methodRef partially applied to the receiver and all the implicit parameters preceding it (A, B, C)
+   *  with the type parameters of the extension (T1, T2) inferred.
+   *  A failure is returned if the implicit search fails for any of the leading implicit parameters or if the receiver has a wrong type
+   *  (note that in general the type of the receiver might depend on the exact types of the found instances of the proceding implicits).
+   *  No implicit search is tried for implicits following the receiver or for parameters of the def (D, E).
+   */
```

Comment:
One question about the specification: if `D` or `E` is not available, maybe we can simply not suggest the extension method to the programmer. This might simplify the code a lot, so that we can simply use `isApplicableMethodRef` as in the original code.

Reply:
As mentioned, `isApplicableMethodRef` doesn't work correctly if there are some leading using sections before the first explicit parameters list. Also as stated in the description of the PR suggesting a method if `D` or `E` are not in scope still makes sense as a user might provide the following implicits explicitly while leading implicits cannot be passed explicitly with dot syntax. This is also consistent with how completions work for implicit classes

Reply:
Maybe do some formatting of the documentation here.